### ISSUE TITLE
BlockClosure>>#isClean regression from Pharo 10 to 11

### ISSUE THREAD

**Issue by koendehondt** [guillep/pharo-core#14141]
**Bug description**
`BlockClosure>>#isClean` does not behave the same on Pharo 11 as it did on Pharo 10.

**To Reproduce**

The problem was found in a much larger block, but we managed to strip it down to the bare essentials to reproduce the bug.

In Pharo 10, this holds:
```smalltalk
[ :x | x someMessageTakingABlock: [ x + 1 ] ] isClean >>> true
```

In Pharo 11, it is different:
```smalltalk
[ :x | x someMessageTakingABlock: [ x + 1 ] ] isClean >>> false
```

**Expected behavior**
I expect the same behaviour in Pharo 10 and Pharo 11.

**Screenshots**
N/A

**Version information:**
 - OS: macOS
 - Version: 13.4
 - Pharo Version: 11

**Expected development cost**
I do not know what the development cost would be, as I am not familiar with the code.
I will be glad to help with testing the fix, although the given example can be turned into a unit test easily.

**Additional context**
None.

**Comment by MarcusDenker** [comment]
Interesting, I will check

**Comment by Ducasse** [comment]
Tx for the report. Sorry for the inconvenience.

**Comment by MarcusDenker** [comment]
I now understand the problem. 

We unified the #isClean to use the one that is defined on the AST, while in Pharo10 we have two implementations: one bytecode level, it just does local analysis of the bc of the block, ignoring embedded blocks.

In Pharo11, we use the AST version, which uses the AST and does a global analysis. Which might be too strict: it never denotes blocks as clean that contain non-clean blocks.

Imagine this:

```
[[ self]  ] isClean 
```
In Pharo10, BC analysis: block is clean, embedded block (the one accessing self) is not clean.

But if we would now compile and execute with clean blocks, the self is a problem: the outer ("clean") block would execute with nil receiver, so self would be nil for the embedded block, too, which would be wrong. Accessing "self" in an embedded block makes all the outer blocks non-clean, because we have to get the "self" from it.

But then, your example, when the internal block does not use self/super (or accesses ivars), I think in that case the outer block could indeed be denoted as clean.


So in Summary: 

- the BC level check in Pharo10 actually was *wrong*, 
- the AST level check that we use from Pharo11 is too conservative, it thinks any block that contains non-clean blocks is not clean, too, while we might be able to relax that notion in the cases the receiver is not needed.

**Comment by MarcusDenker** [comment]
You can test it yourself (the case of just accessing self as above): in RBBlockNode>>#isClean, do not do the super call but return true. 
```
isClean
	"a block is clean if it has no escaping vars, has no non local return and its children are clean"
	self scope hasEscapingVars ifTrue: [ ^ false ].
	self hasBlockReturn ifTrue: [ ^ false ].
	^true
```

Then add a method

```
tt
	^ [[ self] value  ] value  
```

And execute it normally --> returns the instance. Turn on Clean Blocks, recompile the method -> returns nil

**Comment by MarcusDenker** [comment]
(the hasEscapingVars check shows that in case of escaping vars used in the inner block (your example), we already return false at that check, so the experiment is only working for the self case)

**Comment by koendehondt** [comment]
@MarcusDenker Thank you for looking into the issue and thank you for another simple and useful example.

The example I gave is in fact a clean block. The nested block `[ x + 1 ]` is not clean when looking at the block in isolation. Therefore, as you write, the rule "any block that contains non-clean blocks is not clean" is too strict. Lexically speaking `[ :x | x someMessageTakingABlock: [ x + 1 ] ]` is clean. `#isClean` uses a recursive algorithm, which does not take the context into account. In my example, when checking that `[ x + 1 ]` is clean, `#isClean` should be aware of the context, i.e. which variables the block is allowed to access to be clean.

I did not write it down in the initial post, but having the same "isClean" semantics as before is important for my team. We deploy on GemStone, and in order to avoid upgrade issues of our software and the state of the database, we avoid storing non-clean blocks in the database. We use `#isClean` to check that we did not make any mistake in blocks that are supposed to be clean. The current implementation in Pharo 11 does not allow to do that anymore. So we would be very grateful if you could adapt Pharo 11 to consider blocks as given in my example to be clean.

**Comment by MarcusDenker** [comment]
Yes, I want to fix #isClean to allow your example in Pharo11.

I even want to fix Pharo10 to be correct for [[self]] (that this is not clean), as this could lead to blocks being stored without self with Fuel.

**Comment by koendehondt** [comment]
Great Thank you!

**Comment by guillep** [comment]
But `[[self]]` is clean!

I can execute `[[self]] value` and that works ok!

The thing is that it will return a block that is not clean, but that's another issue.

**Comment by MarcusDenker** [comment]
Attached is a .cs of the PR https://github.com/pharo-project/pharo/pull/14282

This way you can easily load it in Pharo11


[isClean.cs.zip](https://github.com/pharo-project/pharo/files/12037724/isClean.cs.zip)


The #isClean is now implemented as:

```
isClean
	"a block is clean if..."
	"it or any embedded blocks do not need self"
	self isAccessingSelf ifTrue: [ ^ false ].
	"it or any embedded blocks do not need the outer context to return"
	self hasNonLocalReturn ifTrue: [ ^ false ].
	"there are no escaping vars accessed from the outer"
	self scope hasEscapingVars ifTrue: [ ^ false ].
	^true
```

**Comment by MarcusDenker** [comment]
I want to do some more experiments and then merge it first in Pharo12, after backport to Pharo11

**Comment by MarcusDenker** [comment]
Intersting numbers: for current Pharo12, we have 10199 clean blocks. With the new definition, we have 153 more.

```
allBlocks := Smalltalk globals methods flatCollect: [:method | method ast blockNodes ].
allBlocks size. "86838"

currentFullBlocks := allBlocks select: [:blockNode | blockNode isInlined not].
currentFullBlocks size.  "36698"

cleanBlocks := currentFullBlocks select: [:blockNode | blockNode isClean].
cleanBlocks size. "10199"

"now file in"

cleanBlocks2 := currentFullBlocks select: [:blockNode | blockNode isClean].
cleanBlocks2 size/ "10352"


(cleanBlocks2 asIdentitySet difference: cleanBlocks asIdentitySet)  "153"


```

**Comment by koendehondt** [comment]
I confirm that the changes in this PR fix the problem I posted. I loaded the changes in a Pharo 11 image to verify. I look forward to the backport of the changes to Pharo 11.

Thank you for the fix.

**Comment by MarcusDenker** [comment]
Backport is merged

**Comment by koendehondt** [comment]
Great. Thanks @MarcusDenker.
